src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,487 @@
+            {
+            // NOTE: This is not asynchronous: it encrypts the data synchronously.
+            // Only writing on the socket is asynchronous.
+            for(auto it = boost::asio::buffer_sequence_begin(buffers);
+                  sent < MAX_PLAINTEXT_SIZE && it != boost::asio::buffer_sequence_end(buffers);
+                  it++)
+               {
+               const std::size_t to_send =
+                  std::min<std::size_t>(MAX_PLAINTEXT_SIZE - sent, boost::asio::buffer_size(*it));
+               native_handle()->send(static_cast<const uint8_t*>(it->data()), to_send);
+               sent += to_send;
+               }
+            }
+         catch(...)
+            {
+            init.completion_handler(Botan::TLS::convertException(), 0);
+            return init.result.get();
+            }
+
+         auto op = create_async_write_op(std::move(init.completion_handler), sent);
+
+         boost::asio::async_write(m_nextLayer, this->m_core.sendBuffer(), std::move(op));
+         return init.result.get();
+         }
+
+      template <typename MutableBufferSequence, typename ReadHandler>
+      BOOST_ASIO_INITFN_RESULT_TYPE(ReadHandler,
+                                    void(boost::system::error_code, std::size_t))
+      async_read_some(const MutableBufferSequence& buffers, ReadHandler&& handler)
+         {
+         BOOST_ASIO_READ_HANDLER_CHECK(ReadHandler, handler) type_check;
+
+         boost::asio::async_completion<ReadHandler,
+               void(boost::system::error_code, std::size_t)>
+               init(handler);
+
+         auto op = create_async_read_op(std::move(init.completion_handler), buffers);
+         op(boost::system::error_code{}, 0);
+         return init.result.get();
+         }
+
+   protected:
+      size_t writePendingTlsData(boost::system::error_code& ec)
+         {
+         auto writtenBytes =
+            boost::asio::write(m_nextLayer, this->m_core.sendBuffer(), ec);
+
+         this->m_core.consumeSendBuffer(writtenBytes);
+         return writtenBytes;
+         }
+
+      template <typename Handler>
+      Botan::TLS::AsyncHandshakeOperation<Channel, StreamLayer, Handler>
+      create_async_handshake_op(Handler&& handler)
+         {
+         return Botan::TLS::AsyncHandshakeOperation<Channel, StreamLayer, Handler>(
+                   native_handle(), this->m_core, m_nextLayer, std::forward<Handler>(handler));
+         }
+
+      template <typename Handler, typename MutableBufferSequence>
+      Botan::TLS::AsyncReadOperation<Channel, StreamLayer, Handler,
+            MutableBufferSequence>
+            create_async_read_op(Handler&& handler,
+                                 const MutableBufferSequence& buffers)
+         {
+         return Botan::TLS::AsyncReadOperation<Channel, StreamLayer, Handler,
+                MutableBufferSequence>(
+                   native_handle(), this->m_core, m_nextLayer, std::forward<Handler>(handler),
+                   buffers);
```

Comment:
These helpers are kind of pointless to be honest, I would just inline them at the call site. Less code and easier to read.